source/StartConditionsPanel.cpp

```diff
@@ -0,0 +1,275 @@
+		double alpha = 1;
+		
+		if(it == chosenStart){
+			FillShader::Fill(zone.Center(), zone.Dimensions(), Color(.1 * alpha, 0.));
+		}
+		
+		string name = font.Truncate(it.GetName(), entryInternalBox.Width());
+		font.Draw(name, point, Color((isHighlighted ? .7 : .5) * alpha, 0.));
+		point += Point(0., entryBox.Height());
+	}
+
+	glDisable(GL_SCISSOR_TEST);
+}
+
+
+
+bool StartConditionsPanel::Drag(double dx, double dy)
+{
+	int x = hoverPoint.X();
+	if(x >= entryListBox.Left() && x < entryListBox.Right())
+	{
+		// Scroll the list
+		// This looks inefficient but it probably gets optimized by the compiler
+		listScroll -= dy;
+		listScroll = max(-entryBox.Height() * (GameData::Start().size()-1), listScroll); // Avoid people going too low
+		listScroll = min(0.,listScroll); // Snap the list to avoid people scrolling too far up
+	}
+	else
+	{
+		// Scroll the description
+		descriptionScroll += dy;
+		descriptionScroll = min(0., descriptionScroll); // Snap the list to avoid people scrolling too far up
+	}
+	
+	return true;
+}
+
+
+
+
+bool StartConditionsPanel::Scroll(double dx, double dy)
+{
+	return Drag(0., dy * Preferences::ScrollSpeed());
+}
+
+
+bool StartConditionsPanel::Hover(int x, int y)
+{
+	hoverPoint = Point(x,y);
+	return true;
+}
+
+bool StartConditionsPanel::KeyDown(SDL_Keycode key, Uint16 mod, const Command &command, bool isNewPress)
+{
+	if(key == 'b')
+	{
+		GetUI()->Pop(this);
+		return true;
+	}
+	else if(key == 'b' || key == SDLK_ESCAPE || command.Has(Command::MENU) || (key == 'w' && (mod & (KMOD_CTRL | KMOD_GUI))))
+		GetUI()->Pop(this);
+	else if(key == 's' || key == 'n' || key == '\n')
+	{
+		if(!hasChosenStart)
+			return true;
+
+		player.New(chosenStart);
+
+
+		ConversationPanel *panel = new ConversationPanel(
+			player, *GameData::Conversations().Get("intro"));
+		GetUI()->Push(panel);
+		panel->SetCallback(this, &StartConditionsPanel::OnCallback);
+	}
+	else 
+		return true;
+	return false;
+}
+
+
+
+bool StartConditionsPanel::Click(int x, int y, int clicks)
+{
+	// The first row of each panel is y = -160 to -140.
+	if(y < -160 || y >= (-160 + 14 * 20))
+		return false;
+	
+
+	if(x >= -470 && x < -250)
+	{
+		int selected = (y + 0 - -160) / 20;
+		int i = 0;
+		for(const auto &it : GameData::Start())
+			if(i++ == selected)
+			{
+				if(!(chosenStart == it))
+					descriptionScroll = 0; // Reset scrolling
+				chosenStart = it;
+				hasChosenStart = true;
+			}
+	}
+	
+	return true;
+}
+
+// Called when the conversation ends
+void StartConditionsPanel::OnCallback(int)
+{
+
+	gamePanels.Reset();
+	gamePanels.CanSave(true);
+	gamePanels.Push(new MainPanel(player));
+	// Tell the main panel to re-draw itself (and pop up the planet panel).
+	gamePanels.StepAll();
+	// If the starting conditions don't specify any ships, let the player buy one.
+	if(player.Ships().empty())
+	{
+		gamePanels.Push(new ShipyardPanel(player));
+		gamePanels.StepAll();
+	}
```

Comment:
This specific code is now repeated in 3 places. We should avoid that. I am OK with that being done in a separate PR, though.